File: DP/64.cc
/**
 * @name 64.最小路径和
 * @details
 * @url https://leetcode.cn/problems/minimum-path-sum/description/?envType=study-plan-v2&envId=dynamic-programming
 * @idea
 * @note
 */
#include "../header.hpp"

class Solution
{
public:
    int minPathSum(vector<vector<int>> &grid)
    {
        int m = grid.size(), n = grid[0].size();
        if (m == 0 || n == 0)
        {
            return 0;
        }
        vector<vector<int>> dp(m, vector<int>(n));
        dp[0][0] = grid[0][0];
        for (int i = 1; i < m; i++)
            dp[i][0] = dp[i - 1][0] + grid[i][0];
        for (int j = 1; j < n; j++)
            dp[0][j] = dp[0][j - 1] + grid[0][j];

        for (int i = 1; i < m; i++)
        {
            for (int j = 1; j < n; j++)
            {
                dp[i][j] = min(dp[i - 1][j], dp[i][j - 1]) + grid[i][j];
            }
        }
        return dp[m - 1][n - 1];
    }
};

int main(int argc, char const *argv[])
{
    Solution s;
}

File: DP/64.py
from typing import List
import numpy as np

class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        # dp = [[0] * len(grid) for _ in len(grid[0])]
        row = len(grid)
        col = len(grid[0])
        dp = np.zeros((row,col))
        dp[0][0] = grid[0][0]
        for i in range (1,row):
            dp[i][0] = dp[i-1][0] + grid[i][0]
        for j in range (1,col):
            dp[0][j] = dp[0][j-1] + grid[0][j]
        for i in range(1,row):
            for j in range(1,col):
                dp[i][j] = min(dp[i-1,j],dp[i,j-1]) + grid[i][j]
                
        return int(dp[row-1][col-1])
            
